rest-lib/src/main/scala/com/gu/mediaservice/lib/auth/Authorisation.scala

```diff
@@ -0,0 +1,51 @@
+package com.gu.mediaservice.lib.auth
+
+import com.gu.mediaservice.lib.argo.ArgoHelpers
+import com.gu.mediaservice.lib.auth.Authentication.{MachinePrincipal, Principal, Request}
+import com.gu.mediaservice.lib.auth.Permissions.{PrincipalFilter, VisibilityFilter}
+import com.gu.mediaservice.lib.auth.provider.AuthorisationProvider
+import play.api.libs.json.Json
+import play.api.mvc.{ActionFilter, Result, Results}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class Authorisation(provider: AuthorisationProvider, executionContext: ExecutionContext) extends Results with ArgoHelpers {
+  def actionFilterFor(permission: PermissionContext[_], unauthorisedResult: Result): ActionFilter[Request] = new ActionFilter[Request] {
+    override protected def filter[A](request: Request[A]): Future[Option[Result]] = {
+      if (hasPermissionTo(permission)(request.user)) {
+        Future.successful(None)
+      } else {
+        Future.successful(Some(unauthorisedResult))
+      }
+    }
+    override protected def executionContext: ExecutionContext = Authorisation.this.executionContext
+  }
+  def actionFilterFor(permission: PermissionContext[_]): ActionFilter[Request] =
+    actionFilterFor(
+      permission,
+      respondError(Unauthorized, "permission-denied", s"You do not have permission to ${permission.permission}")
+    )
+
+  def hasPermissionTo[T](permission: SimplePermission): PrincipalFilter =
+    hasPermissionTo(PermissionContext(permission))
+
+  def hasPermissionTo[T](permission: PermissionWithParameter[T], parameter: T): PrincipalFilter =
+    hasPermissionTo(permission -> parameter)
+
+  def hasPermissionTo[T](permission: PermissionContext[T]): PrincipalFilter = {
+    principal: Principal => {
+      principal match {
+        // a machine principal with internal tier can always see anything
+        case MachinePrincipal(ApiAccessor(_, Internal), _) => true
+        case _ => provider.hasPermissionTo(permission, principal)
+      }
+    }
+  }
+
+  def visibilityFilterFor[T](permission: PermissionWithParameter[T], principal: Principal): VisibilityFilter[T] =
```

Comment:
Hello @sihil.
We are wondering if this method would work for filtering images by permission, i.e., showing to user a set of images depending on his permission.

Reply:
That's the intention. Obviously parameterised permissions are a little down the road and this probably won't be entirely fit for purpose so I expect at the time it will need to be refined. In particular I imagine that we might sometimes want to shape an ES query from a set of `T` that a user can see, where `T` might be organisation.

Reply:
I've stripped out the more complex permissions model for the time being. We can perhaps arrange for the BBC/Guardian to pair on it once we have a concrete use case. I've deliberately left it as an unsquashed commit so it is there if we want it back.